ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/internal/methods/DebugSetHead.java

```diff
@@ -45,26 +64,94 @@ public String getName() {
   }
 
   @Override
-  protected BlockParameter blockParameter(final JsonRpcRequestContext request) {
+  protected BlockParameterOrBlockHash blockParameterOrBlockHash(
+      final JsonRpcRequestContext requestContext) {
     try {
-      return request.getRequiredParameter(0, BlockParameter.class);
+      return requestContext.getRequiredParameter(0, BlockParameterOrBlockHash.class);
     } catch (JsonRpcParameterException e) {
       throw new InvalidJsonRpcParameters(
-          "Invalid block parameter (index 0)", RpcErrorType.INVALID_BLOCK_PARAMS, e);
+          "Invalid block or block hash parameter (index 0)", RpcErrorType.INVALID_BLOCK_PARAMS, e);
     }
   }
 
   @Override
-  protected Object resultByBlockNumber(
-      final JsonRpcRequestContext request, final long blockNumber) {
-    final Optional<Hash> maybeBlockHash = getBlockchainQueries().getBlockHashByNumber(blockNumber);
+  protected Object resultByBlockHash(final JsonRpcRequestContext request, final Hash blockHash) {
+    var blockchainQueries = getBlockchainQueries();
+    var blockchain = protocolContext.getBlockchain();
+    Optional<BlockHeader> maybeBlockHeader = blockchainQueries.getBlockHeaderByHash(blockHash);
+    Optional<Boolean> maybeMoveWorldstate = shouldMoveWorldstate(request);
 
-    if (maybeBlockHash.isEmpty()) {
+    if (maybeBlockHeader.isEmpty()) {
       return new JsonRpcErrorResponse(request.getRequest().getId(), UNKNOWN_BLOCK);
     }
 
-    protocolContext.getBlockchain().rewindToBlock(maybeBlockHash.get());
+    // Optionally move the worldstate to the specified blockhash, if it is present in the chain
+    if (maybeMoveWorldstate.orElse(Boolean.FALSE)) {
+      var archive = blockchainQueries.getWorldStateArchive();
+
+      // Only DiffBasedWorldState's need to be moved:
+      if (archive instanceof DiffBasedWorldStateProvider diffBasedArchive) {
+        rollWorldStateIncrementally(maybeBlockHeader.get(), blockchain, diffBasedArchive);
+      }
+    }
+
+    // finally move the blockchain
+    blockchain.rewindToBlock(maybeBlockHeader.get().getBlockHash());
```

Comment:
The implication of moving the chain after moving the state is that we will not be able to move the worldstate FORWARD more than 512 blocks in the same call as the blockchain.  It would have to be split up into 2 calls, move the chain without worldstate, then move the chain with worldstate.

This is just due to the nature of doing the incremental state rolling.  The canonical chain needs to be at the target hash before we can do the incremental rolling.  Since there could be forks, and there isn't a canonical block number without the chain having moved first, we have to roll the chain forward first.   

We could flip this consideration and allow for arbitrary incremental forward rolling by moving the blockchain first, but that just flips the problem around such that we cannot move incrementally backward.  

This is less than ideal, but it is worth it to enable us to safely move the worldstate without exhausting the memory in an OOM and/or db transaction memory for long state moves.

if we move the chain by hash 512 or fewer blocks at a time, this is not a concern.

Reply:
thinking about it today, we should be able to move the chain in the same increments as the worldstate and make this not a problem.  If they move in-tandem, then we should always be able to resolve where we are moving the worldstate to, forward or backward.

Reply:
added chain and worldstate incremental move.